core/src/main/java/io/parsingdata/metal/ImmutableObject.java

```diff
@@ -0,0 +1,26 @@
+package io.parsingdata.metal;
+
+/**
+ * When objects are immutable, that means their hashcode will stay the same the moment it is created.
+ * It will improve performance if these objects cache their hash.
+ * <p>
+ * This is a lazy implementation, instead of calculating the hash once within the constructor, to avoid
+ * performance decrease during parsing. In most implementations the hash is not actually used/needed.
+ */
+public abstract class ImmutableObject {
+
+    private Integer hash;
```

Comment:
Another consideration: This is an `Integer`, and not an `int`, because it can be `null`.

Using boxed primitives isn't "free", though. The `int` would be 4 bytes, but each boxed `Integer` is 16 or even 24 bytes (according to [this Stackoverflow answer](https://stackoverflow.com/a/9037808)).

One alternative here could be to have a primitive `int hashCode` and a `boolean initialized`.

Not sure if this is worth the effort, but since most classes now inherit from this one, and they are used all over the `ParseGraph`, it might be worth considering.

This would be an easy fix in the future, so for now I think it's okay to keep this current implementation.

Reply:
Ah, interesting point! I found the use of an object here quite elegant, opposed to a separate boolean, but I haven't considered how much memory this would need. There are two concerns here:
1. Performance
2. Memory usage

1. Performance: I have run the performance test for large numbers and there does not seem to be a difference when using int over Integer. I think the JVM does a lot of auto improvements here.
2. Memory usage: hashCodes are not calculated for all objects. They are only calculated when needed. Since they are initialized as null, they only require as much space as a pointer requires, which is 4 bytes like a primitive int. When we would use ints + booleans as fields, then this requires a little more memory to store the initialize boolean variable. Assuming a boolean requires a bit of space (i mean this literal ;)), and a Integer hashCode requires 160 bits additional bits when calculated, then that means that if we calculate less then 1/160th of the total number of ImmutableObjects, then the Integer implementation actually requires less memory.


Tl;dr;
So, in short, performance is not an issue and since memory usage depends on the implementation on how many objects you want to calculates hashes for (and the default is for 0 objects), it may be wise to keep this as is.